Sources/Data Model/ProjectConfig.swift

```diff
@@ -171,32 +167,40 @@ class ProjectConfig {
         
     }
     
-    private func assignHoldoutIdsToFeatureFlags() {
-        let flagsWithHoldoutIds = project.featureFlags.map { flag -> FeatureFlag in
-            var updatedFlag = flag
-            var holdoutIds = [String]()
-            for holdout in project.holdouts {
-                if !holdout.includedFlags.isEmpty {
-                    if holdout.includedFlags.contains(flag.id) {
+    func getHoldoutIdsForFlag(id: String) -> [String] {
+        guard !project.holdouts.isEmpty else { return [] }
+        
+        if let holdoutIds = flagHoldoutsMap[id] {
+            return holdoutIds
+        }
+        
+        updateHoldoutsMapForFlag(id: id)
+        
+        return flagHoldoutsMap[id] ?? []
+    }
+    
+    private func updateHoldoutsMapForFlag(id: String) {
+        var holdoutIds = [String]()
+        
+        for holdout in project.holdouts {
+            switch (holdout.includedFlags.isEmpty, holdout.excludedFlags.isEmpty) {
+                case (true, true):
+                    // Global holdout
+                    holdoutIds.append(holdout.id)
+                    
+                case (false, _):
+                    if holdout.includedFlags.contains(id) {
                         holdoutIds.append(holdout.id)
                     }
-                } else if !holdout.excludedFlags.isEmpty {
-                    if !holdout.excludedFlags.contains(flag.id) {
+                    
+                case (_, false):
+                    if !holdout.excludedFlags.contains(id) {
                         holdoutIds.append(holdout.id)
                     }
```

Comment:
A couple of suggestions - 
1. how can we force the evaluation ordering: globals first (per TDD) and then we should add local (included).
2. We can build map of HOs per flag-ids when project building, so we can access it for decision -
includesHOs = {
  flag_id: [ ho1, ho2 ],
  ...
} 
excludesHOs = {
  flag_id: [ ho1 ],
  ...
}
otherHOs = [ ho3, ho4, ... ]